libc/src/__support/FPUtil/BasicOperations.h

```diff
@@ -178,6 +181,66 @@ LIBC_INLINE T fdim(T x, T y) {
   return (x > y ? x - y : 0);
 }
 
+template <typename T, cpp::enable_if_t<cpp::is_floating_point_v<T>, int> = 0>
+LIBC_INLINE int canonicalize(T &cx, const T &x) {
+  FPBits<T> sx(x);
+  if constexpr (get_fp_type<T>() == FPType::X86_Binary80) {
+    // All the pseudo and unnormal numbers are not canonical.
+    // More precisely :
+    // Exponent   |       Significand      | Meaning
+    //            | Bits 63-62 | Bits 61-0 |
+    // All Ones   |     00     |    Zero   | Pseudo Infinity, Value = Infinty
```

Comment:
This is not correct.

The x86 manual states:
> Specifically, the categories of encodings formerly known as pseudo-NaNs, pseudo-infinities, and un-normal
numbers are not supported and should not be used as operand values. The Intel 387 math coprocessor and later
IA-32 processors generate an invalid-operation exception when these encodings are encountered as operands.

So pseudo-infinity and pseudo-denormal should be canonicalized to sNaN, not infinity or finite value (respectively).

Reply:
Ah I see, from the wikipedia page : 

> Pseudo-Infinity. The sign bit gives the sign of the infinity. The 8087 and 80287 treat this as Infinity. The 80387 and later treat this as an invalid operand.

I focused only the 8087 and 80287 part. Will link a patch soon.

Reply:
@jcranmer-intel : do you mean to treat them as `sNaN`, not literally return `sNaN`?  B/c `canonicalize` translates `sNaN` to `qNaN + FE_INVALID`.

Reply:
@jcranmer-intel, can you confirm the above.

Reply:
Yes, they should be treated as sNaN, and therefore returned as qNaN

(alternatively, use the escape hatch of don't-canonicalize-this-value that C provides--glibc takes this approach).